File: bindings/wasm/audit_trail_wasm/examples/src/04_configure_locking.ts
// Copyright 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

/**
 * ## Actors
 *
 * - **Admin client**: Creates the trail and sets up the LockingAdmin and RecordAdmin roles.
 * - **Locking admin client**: Controls write and delete locks. Holds the LockingAdmin capability.
 * - **Record admin client**: Writes records. Used to demonstrate that the write lock is enforced
 *   per-sender, not just checked by the admin client.
 *
 * Demonstrates how to:
 * 1. Delegate locking updates through a LockingAdmin role.
 * 2. Freeze record creation with a write lock.
 * 3. Restore writes and add a new record.
 * 4. Update the delete-record window and delete-trail lock.
 */

import {
    CapabilityIssueOptions,
    Data,
    LockingConfig,
    LockingWindow,
    PermissionSet,
    TimeLock,
} from "@iota/audit-trail/node";
import { strict as assert } from "assert";
import { createTrailWithSeedRecord, getFundedClient, TEST_GAS_BUDGET } from "./util";

export async function configureLocking(): Promise<void> {
    console.log("=== Audit Trail: Configure Locking ===\n");

    // `adminClient` creates the trail and delegates separate lock/write authority.
    // `lockingAdminClient` controls locks; `recordAdminClient` writes records.
    const adminClient = await getFundedClient();
    const lockingAdminClient = await getFundedClient();
    const recordAdminClient = await getFundedClient();

    const { output: createdTrail } = await createTrailWithSeedRecord(adminClient);
    const trailId = createdTrail.id;

    const lockingAdminRole = adminClient.trail(trailId).access().forRole("LockingAdmin");
    await lockingAdminRole
        .create(PermissionSet.lockingAdminPermissions())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);
    await lockingAdminRole
        .issueCapability(new CapabilityIssueOptions(lockingAdminClient.senderAddress()))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    const recordAdminRole = adminClient.trail(trailId).access().forRole("RecordAdmin");
    await recordAdminRole
        .create(PermissionSet.recordAdminPermissions())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);
    await recordAdminRole
        .issueCapability(new CapabilityIssueOptions(recordAdminClient.senderAddress()))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(adminClient);

    // LockingAdmin freezes writes.
    await lockingAdminClient
        .trail(trailId)
        .locking()
        .updateWriteLock(TimeLock.withInfinite())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(lockingAdminClient);

    const lockedTrail = await adminClient.trail(trailId).get();
    console.log("Write lock after update:", lockedTrail.lockingConfig.writeLock, "\n");
    assert.equal(lockedTrail.lockingConfig.writeLock.type, TimeLock.withInfinite().type);

    // RecordAdmin attempts to add a record while locked — should fail.
    const blockedAdd = await recordAdminClient
        .trail(trailId)
        .records()
        .add(Data.fromString("This write should fail"), "blocked")
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(recordAdminClient)
        .catch(() => null);
    assert.equal(blockedAdd, null, "write lock should block adding records");

    // LockingAdmin lifts the write lock.
    await lockingAdminClient
        .trail(trailId)
        .locking()
        .updateWriteLock(TimeLock.withNone())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(lockingAdminClient);

    const recordAddedAfterUnlock = await recordAdminClient
        .trail(trailId)
        .records()
        .add(Data.fromString("Write lock lifted"), "event:resumed")
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(recordAdminClient);
    console.log("Added record", recordAddedAfterUnlock.output.sequenceNumber, "after clearing the write lock.\n");

    // LockingAdmin configures deletion window and trail lock.
    await lockingAdminClient
        .trail(trailId)
        .locking()
        .updateDeleteRecordWindow(LockingWindow.withCountBased(BigInt(2)))
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(lockingAdminClient);
    await lockingAdminClient
        .trail(trailId)
        .locking()
        .updateDeleteTrailLock(TimeLock.withInfinite())
        .withGasBudget(TEST_GAS_BUDGET)
        .buildAndExecute(lockingAdminClient);

    const finalTrail = await adminClient.trail(trailId).get();
    console.log("Final locking config:");
    console.log("  delete_record_window =", finalTrail.lockingConfig.deleteRecordWindow);
    console.log("  delete_trail_lock =", finalTrail.lockingConfig.deleteTrailLock);
    console.log("  write_lock =", finalTrail.lockingConfig.writeLock);

    assert.equal(finalTrail.lockingConfig.deleteRecordWindow.type, LockingWindow.withCountBased(BigInt(2)).type);
    assert.equal(finalTrail.lockingConfig.deleteTrailLock.type, TimeLock.withInfinite().type);
    assert.equal(finalTrail.lockingConfig.writeLock.type, TimeLock.withNone().type);
}

File: examples/audit-trail/04_configure_locking.rs
// Copyright 2020-2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! ## Actors
//!
//! - **Admin client**: Creates the trail and sets up the LockingAdmin and RecordAdmin roles.
//! - **Locking admin client**: Controls write and delete locks. Holds the LockingAdmin capability.
//! - **Record admin client**: Writes records. Used to demonstrate that the write lock is enforced per-sender, not just
//!   checked by the admin.

use anyhow::{Result, ensure};
use audit_trail::core::types::{CapabilityIssueOptions, Data, InitialRecord, LockingWindow, PermissionSet, TimeLock};
use examples::get_funded_audit_trail_client;
use product_common::core_client::CoreClient;

/// Demonstrates how to:
/// 1. Delegate locking updates through a `LockingAdmin` role.
/// 2. Freeze record creation with a write lock.
/// 3. Restore writes and add a new record.
/// 4. Update the delete-record window and delete-trail lock.
#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Audit Trail: Configure Locking ===\n");

    // Use separate clients to show that locking and record-writing permissions can be delegated independently.
    let admin_client = get_funded_audit_trail_client().await?;
    let locking_admin_client = get_funded_audit_trail_client().await?;
    let record_admin_client = get_funded_audit_trail_client().await?;

    let created_trail = admin_client
        .create_trail()
        .with_initial_record(InitialRecord::new(
            Data::text("Trail opened"),
            Some("event:created".to_string()),
            None,
        ))
        .finish()
        .build_and_execute(&admin_client)
        .await?
        .output;

    let trail_id = created_trail.trail_id;
    let locking_admin_role = "LockingAdmin";
    let record_admin_role = "RecordAdmin";

    // The Admin capability authorizes defining roles and issuing the delegated capabilities.
    admin_client
        .trail(trail_id)
        .access()
        .for_role(locking_admin_role)
        .create(PermissionSet::locking_admin_permissions(), None)
        .build_and_execute(&admin_client)
        .await?;
    admin_client
        .trail(trail_id)
        .access()
        .for_role(locking_admin_role)
        .issue_capability(CapabilityIssueOptions {
            issued_to: Some(locking_admin_client.sender_address()),
            valid_from_ms: None,
            valid_until_ms: None,
        })
        .build_and_execute(&admin_client)
        .await?;

    admin_client
        .trail(trail_id)
        .access()
        .for_role(record_admin_role)
        .create(PermissionSet::record_admin_permissions(), None)
        .build_and_execute(&admin_client)
        .await?;
    admin_client
        .trail(trail_id)
        .access()
        .for_role(record_admin_role)
        .issue_capability(CapabilityIssueOptions {
            issued_to: Some(record_admin_client.sender_address()),
            valid_from_ms: None,
            valid_until_ms: None,
        })
        .build_and_execute(&admin_client)
        .await?;

    locking_admin_client
        .trail(trail_id)
        .locking()
        .update_write_lock(TimeLock::Infinite)
        .build_and_execute(&locking_admin_client)
        .await?;

    let locked_trail = admin_client.trail(trail_id).get().await?;
    println!(
        "Write lock after update: {:?}\n",
        locked_trail.locking_config.write_lock
    );
    ensure!(locked_trail.locking_config.write_lock == TimeLock::Infinite);

    let blocked_add = record_admin_client
        .trail(trail_id)
        .records()
        .add(Data::text("This write should fail"), None, None)
        .build_and_execute(&record_admin_client)
        .await;
    ensure!(blocked_add.is_err(), "write lock should block adding records");

    locking_admin_client
        .trail(trail_id)
        .locking()
        .update_write_lock(TimeLock::None)
        .build_and_execute(&locking_admin_client)
        .await?;

    let added_record = record_admin_client
        .trail(trail_id)
        .records()
        .add(Data::text("Write lock lifted"), Some("event:resumed".to_string()), None)
        .build_and_execute(&record_admin_client)
        .await?
        .output;

    println!(
        "Added record {} after clearing the write lock.\n",
        added_record.sequence_number
    );

    locking_admin_client
        .trail(trail_id)
        .locking()
        .update_delete_record_window(LockingWindow::CountBased { count: 2 })
        .build_and_execute(&locking_admin_client)
        .await?;
    locking_admin_client
        .trail(trail_id)
        .locking()
        .update_delete_trail_lock(TimeLock::Infinite)
        .build_and_execute(&locking_admin_client)
        .await?;

    let final_trail = admin_client.trail(trail_id).get().await?;
    println!(
        "Final locking config:\n  delete_record_window = {:?}\n  delete_trail_lock = {:?}\n  write_lock = {:?}",
        final_trail.locking_config.delete_record_window,
        final_trail.locking_config.delete_trail_lock,
        final_trail.locking_config.write_lock
    );

    ensure!(final_trail.locking_config.delete_record_window == LockingWindow::CountBased { count: 2 });
    ensure!(final_trail.locking_config.delete_trail_lock == TimeLock::Infinite);
    ensure!(final_trail.locking_config.write_lock == TimeLock::None);

    Ok(())
}
